### PR TITLE
Add cancellation token support at boundaries

### PR DESCRIPTION
**Pull request by brockallen**
This PR adds a service to DI that presents the cancellation token from the HTTP context to the boundaries where they might be needed, such as outbound DB calls to EF. 

Fixes: https://github.com/DuendeSoftware/IdentityServer/issues/184

### CODE REVIEW

File: src/IdentityServer/Services/Default/DefaultCancellationTokenService.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+using Microsoft.AspNetCore.Http;
+using System.Threading;
+
+namespace Duende.IdentityServer.Services.Default
+{
+    class DefaultCancellationTokenService : ICancellationTokenService
```
Comment (damianh) [inline src/IdentityServer/Services/Default/DefaultCancellationTokenService.cs:6]: 🤔 Perhaps `HttpContextCancellationTokenService` ? Bit more intention revealing?

**Comment by brockallen** [comment]
@damianh Thanks for the feedback -- beyond that, any other input? As we discussed, it's a half-way solution to the problem given the code base history. Do you feel like it's sufficient for your purposes?

**Comment by damianh** [comment]
Need to get a deeper look at it. Should have just started a review there, hit the "comment now" button. Will get at it shorty.

File: src/Storage/Services/ICancellationTokenService.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+
+using System.Threading;
+
+namespace Duende.IdentityServer.Services
+{
+    /// <summary>
+    /// Service to provide CancellationToken for async operations.
+    /// </summary>
+    public interface ICancellationTokenService
```
Comment (damianh) [inline src/Storage/Services/ICancellationTokenService.cs:12]: I'm thinking perhaps `ICancellationTokenProvider` (not a hill to die on.)

**Review by damianh — APPROVED** [review]
Still can't help the feeling that I'd prefer have it as a method parameter. However this works, no changes to suggest.

**Comment by brockallen** [comment]
> Still can't help the feeling that I'd prefer have it as a method parameter. However this works, no changes to suggest.

Yep, and this is the compromise to not breaking almost every interface/service we have defined. If we had it to do over again, I'd have been 100% in agreement with your original proposal. If at some point in the future we have a "break the world" release, we'd do it then (but I don't if that will ever happen).

**Comment by brockallen** [comment]
And thank you for the feedback and time on this!

**Pull request merged**